lib/srv/regular/sshserver.go

```diff
@@ -1096,40 +1109,253 @@ func (s *Server) getServerResource() (types.Resource, error) {
-	if d, ok := ctx.Parent().GetDirectTCPIPForwardDialer(); ok {
+func (s *Server) getDirectTCPIPForwardDialer(scx *srv.ServerContext) (sshutils.TCPIPForwardDialer, error) {
+	if d, ok := scx.Parent().GetDirectTCPIPForwardDialer(); ok {
 		return d, nil
 	}
 
-	dialerConn, listenerConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	proc, err := s.startForwardingSubprocess(scx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// set up a dial function that sends the address + fd as a unix datagram message. the forwarder subprocess
+	// interprets all such messages in this way, and will dial the specified address and proxy all traffic
+	// to the desired endpoint.
+	dialer := func(addr string) (net.Conn, error) {
+		local, remote, err := uds.NewSocketpair(uds.SocketTypeStream)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer remote.Close()
+
+		remoteFD, err := remote.File()
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+		defer remoteFD.Close()
+
+		_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		return local, nil
+	}
+
+	// try to register with the parent context.
+	if other, ok := scx.Parent().TrySetDirectTCPIPForwardDialer(dialer); !ok {
+		// another forwarder was concurrently created. this isn't actually a problem, multiple forwarders
+		// being registered is harmless, but it does result in slightly higher resource utilization, so its
+		// preferable to use the existing forwarder and close ours in the background.
+		go proc.Close()
+		return other, nil
+	}
+
+	// successfully registered this dialer, add closer to context.
+	scx.Parent().AddCloser(proc)
+
+	return dialer, nil
+}
+
+// listenTCPIP creates a new listener in the forwarding process.
+func (s *Server) listenTCPIP(scx *srv.ServerContext, addr string) (net.Listener, error) {
+	proc, err := s.getTCPIPForwardProcess(scx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create a socket to receive new connections on.
+	localConn, remoteConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer localConn.Close()
+	defer remoteConn.Close()
+	remoteFD, err := remoteConn.File()
+	if err != nil {
+		localConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	defer remoteFD.Close()
+	_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The forwarding process may have chosen its own port, so we need to get the
+	// new listen address.
+	var fn int
+	fbuf := make([]*os.File, 1)
+	readDone := make(chan struct{})
+	// Read addr in another goroutine so we can cancel it if the forwarding process
+	// stops.
+	go func() {
+		_, fn, err = localConn.ReadWithFDs(nil, fbuf)
+		close(readDone)
+	}()
+
+	select {
+	case <-proc.Done:
+		localConn.Close()
+		return nil, trace.Errorf("forwarding process is closed")
```

Comment:
`fbuf` might be filled by the `ReadWithFDs` after we select this case of the select but before we close `localConn`, so that'd result in the file descriptor leaking until the finalizer (if any) closes it; keep `readDone` as a synchronous channel but make it a `chan *os.File`, and in the goroutine do a select on `readDone<-f` and `proc.Done` so you can close the returned file descriptor (if any) if the parent goroutine doesn't want it anymore.